### ISSUE TITLE
on ACD. large files give 504 error. The files appears to be on ACD. Rclone redoes the upload.

### ISSUE THREAD

**Issue by left1000** [ncw/rclone#262]
on ACD. large files give 504 error. The files appears to be on ACD. Rclone redoes the upload.

This seems to happen roughly 100% of the time on files of size 10-50GB that I've tested it on so far.
The end result of some files is that they upload fine the first try. Some files will 504 three times and then rclone will exit (it only retries the command three times).... after the third time the file is just on ACD... as far as I can tell all three uploads worked perfectly. While the error message appears almost always on these files the unneeded repetitions only happens maybe 10-20% of the time? or looking over my test files it seems the unneeded repetitions happen always on files greater than 31gb and on files 10-31gb the error message appears but then in the second round the check realizes it doesn't need to repeat and so doesn't.

I'm on ethernet not wireless which is why I didn't just post in this thread: https://github.com/ncw/rclone/issues/183

edit: here's a quote of an exact error of this type:
2015/12/24 20:59:52 gigatribe stuff 1.5.7z.003: Failed to copy: HTTP code 504: "
504 GATEWAY_TIMEOUT", no response body

**Comment by ncw** [comment]
There is a problem with downloading > 10GB files with ACD  #204 which I've worked out what is causing it, but I haven't had time to fix it yet.

This looks like it isn't related though.

What is the command line you are using?  Are you using `--bwlimit`?  If so #232 might be the cause.  It certainly looks similar though.

**Comment by left1000** [comment]
now that I think of it... I have both used --bwlimit AND not used --bwlimit.... when NOT using --bwlimit I got all the same 504 errors BUT each file only uploaded once... I do believe some of the files which were uploaded 2or3 times in a row (with each copy being presumably perfect) were with --bwlimit.

I'll generate a -v of one of these situations later, these huge files take me a long time to upload so I guess getting to the -v of the situation will take quite a while.

**Comment by ncw** [comment]
I have uploaded a new version of rclone a moment ago which definitely has the bandwidth limit fixes for amazon cloud drive.

**Comment by ncw** [comment]
I made a thread on the amazon developer forum here: https://forums.developer.amazon.com/forums/thread.jspa?threadID=10979

A log of this happening with `--dump-headers` would be very useful.  Remove the `Authorization:` header before posting it publically.

**Comment by santaclause88** [comment]
I'm also seeing this problem with large files >20gb ish. Using rclone-v1.29-linux-amd64

**Comment by ncw** [comment]
@santaclause88 if you can get me a log of it happening I can post it on the thread which might get some traction from amazon!

> A log of this happening with `--dump-headers` would be very useful. Remove the `Authorization:` header before posting it publically.

**Comment by santaclause88** [comment]
Here is a log. It has been trimmed up a bit to save some space. Let me know how else I can help. 
[log.txt](https://github.com/ncw/rclone/files/250454/log.txt)

**Comment by ncw** [comment]
@santaclause88 Thanks for that. I've given the forum thread some info from the log - lets hope Amazon take a look.

**Comment by DurvalMenezes** [comment]
I'm getting the same error here, local is Linux Devuan 1.0 (systemd-less Debian 8)  running rclone v.133, and remote is an *encrypted* Amazon Cloud Drive (using an rclone "type = crypt" section, configured on top of an Amazon Drive section). 

I'm confused: I was under the impression that the size limit on Amazon Cloud Drive was 50GB? I'm getting issues with files exactly 45GB (ie, 45000000000 bytes), but apparently in a random manner (sometimes they go through, other times rclone (incorrectly) assumes they are not there and retries them from the start, even as an "rclone ls" in parallel shows the files *are* there, for example:

This is the "-v" log:
================================================================[
 * TESTFILE_ac: 100% done. avg: 2528.5, cur:   16.8 kByte/s. ETA: 0s
 2016/09/30 19:52:24 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (2/24)
 2016/09/30 19:52:29 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (3/24)
 2016/09/30 19:52:34 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (4/24)
 2016/09/30 19:52:41 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (5/24)
 2016/09/30 19:52:47 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (6/24)
 2016/09/30 19:52:52 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (7/24)
 2016/09/30 19:52:57 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (8/24)
 2016/09/30 19:53:02 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (9/24)
 2016/09/30 19:53:08 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (10/24)
 2016/09/30 19:53:13 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (11/24)
 2016/09/30 19:53:18 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (12/24)
 2016/09/30 19:53:20
[...]
 * TESTFILE_ac: 100% done. avg: 2519.8, cur:    0.3 kByte/s. ETA: 0s
2016/09/30 19:53:23 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (13/24)
2016/09/30 19:53:28 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (14/24)
2016/09/30 19:53:34 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (15/24)
2016/09/30 19:53:39 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (16/24)
2016/09/30 19:53:44 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (17/24)
2016/09/30 19:53:49 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (18/24)
2016/09/30 19:53:55 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (19/24)
2016/09/30 19:54:00 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (20/24)
2016/09/30 19:54:05 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (21/24)
2016/09/30 19:54:11 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (22/24)
2016/09/30 19:54:16 &{TESTFILE_ac 0xc420298a20}: Object not found - waiting (23/24)                                                                           2016/09/30 19:54:19 &{TESTFILE_aa 0xc420298a20}: Error detected after finished upload - waiting to see if object was uploaded correctly: HTTP code 504: "504 GATEWAY_TIMEOUT", no response body
2016/09/30 19:54:26 pacer: Rate limited, sleeping for 718.450071ms (1 consecutive low level retries)
2016/09/30 19:54:26 pacer: low level retry 1/1 (error HTTP code 504: "504 GATEWAY_TIMEOUT", no response body
)                                                                                                                                                             2016/09/30 19:54:26 TESTFILE_ac: Received error: HTTP code 504: "504 GATEWAY_TIMEOUT", no response body
 - low level retry 2/10
 2016/09/30 19:54:26 pacer: Resetting sleep to minimum 20ms on success
[...]
Transferring:
 * TESTFILE_ac:  0% done. avg: 5836.3, cur: 6596.3 kByte/s. ETA: 1h50m14s
================================================================]

This is the output of "rclone ls amzcdcrypt:destdir:
================================================================[
 45000000000 TESTFILE_ac
2016/09/30 20:13:38 
Transferred:      0 Bytes (0 Bytes/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:        2.4s
================================================================]

So, what should I do? Limit files to under 10GB?

Thanks, 
 -- 
   Durval.

**Comment by felixbuenemann** [comment]
I have also seen this happening in the official amazon client for the mac: It uploads to 100%, then it gets a 504 gateway timeout and immediately restarts the upload. If you cancel the re-upload and look into drive, you will see that the file is already there.

So I think in the 504 case at 100% the behavior should be handled like the successful upload case and it should wait for acd-wait-time for the file to appear.

**Comment by arionl** [comment]
I'm seeing the same issue as reported. I was running with '-v' but not debug. Here's the tail of my log file if it is any help. While rclone was doing it's 24 retries I was also checking ACD with the web interface and didn't see the file.. however, when rclone started sending the file again from scratch, the file popped up in ACD.. not sure if that was just coincidence -- i.e. just took a little longer for the file to show up -- or if the retry triggered the file showing up..

[large-upload-error.txt](https://github.com/ncw/rclone/files/518531/large-upload-error.txt)

**Comment by ncw** [comment]
Try Increasing --acd-upload-wait-time the default is 2m. It maybe needs to scale as the Suze of file. How big are the files you are uploading?

**Comment by arionl** [comment]
The file is ~35gb. I'll try uploading the same file again with a ```--acd-upload-wait-time``` of 15 mins and see if that makes a difference and report back..

**Comment by arionl** [comment]
I tried again and it looks like it took ~5mins for the file to be "seen":

```
2016/10/09 21:30:01 video.m2ts: Object not found - waiting (40/180)
2016/10/09 21:30:06 video.m2ts: Object not found - waiting (41/180)
2016/10/09 21:30:11 video.m2ts: Object not found - waiting (42/180)
2016/10/09 21:30:16 video.m2ts: Object not found - waiting (43/180)
2016/10/09 21:30:21 video.m2ts: Object found with correct size - returning with no error
2016/10/09 21:30:21 video.m2ts: Copied (new)
2016/10/09 21:30:21
Transferred:   32.773 GBytes (3.282 MBytes/s)
Errors:                 0
Checks:                 0
Transferred:            1
Elapsed time:  2h50m25.2s
2016/10/09 21:30:21 Go routines at exit 4
```

Anyone else seeing similar behavior? Perhaps an increase in default from 2 mins may be warranted?

**Comment by felixbuenemann** [comment]
I think there's a relation between the size of the file and the time it takes to show up. Maybe rclone could add a log message that logs the relation of file size to how long it took for the file to appear. After gathering that data for some time it could be used to derive a common formula for the default wait time. Something like x-seconds + (m-seconds/gigabyte * n-gigabytes).

**Comment by ncw** [comment]
@felixbuenemann good idea.  I'll put that in then we can do a bit of testing.

**Comment by felixbuenemann** [comment]
@ncw Btw. I noticed that rclone also re-uploads a file if it gets rate-limited at the end of the upload, because it doesn't jump into the wait-for-upload code in that case, so even though the file uploaded fine it is re-uploaded anyways.

Should I open a new issue for that case?

**Comment by ncw** [comment]
@felixbuenemann I've fixed that already - try the latest beta: http://beta.rclone.org

**Comment by felixbuenemann** [comment]
@ncw That is the build I am currently running, but there is a chance the copy process where the problem happened was still running an older build, because I didn't restart it after updating. I've rebooted my system since then so I can say for sure. I'll open a new issue if it happens again.

**Comment by ncw** [comment]
@felixbuenemann thanks - the fix was #712 BTW

**Comment by ncw** [comment]
I've done a bit of analysis.  Firstly there is a 60 second timeout before we get the `408 GATEWAY_TIMEOUT` message.  You can see this because the stats show the files get to 100% 60 seconds before the error.

Taking that into account, the timing for various sized files look like this (I can only upload files up oto 20GB as that is the limit of my account - I put the 35 GB info from @arionl in too).

![acd-timeouts2](https://cloud.githubusercontent.com/assets/536803/19393938/d39c4538-922e-11e6-99b2-741b0957c6ca.png)

Which graph up like this

![acd-timeouts](https://cloud.githubusercontent.com/assets/536803/19393948/e0c81638-922e-11e6-96e2-309b21dd6ff4.png)

So what that means is that amazon take about 8.2 seconds per GB hashing the file == 125 MByte/s.  If that time gets to be greater than 60s then we get a `408 GATEWAY_TIMEOUT` error.

So a prediction of how long we need to wait after a `408 GATEWAY_TIMEOUT` error would be `FILE_SIZE_IN_GB * 8.2 - 60` seconds.

The outlier in the table would have needed 21.7 S/GB though, so to be sure of always waiting long enough we should wait (say) 30 seconds / GB - 60 seconds for the initial timeout.  So for a < 2GB file we wouldn't wait at all, for a 3GB file we would wait for 30s, for a 10GB file we would wait for 4mins30 and a 50 GB file we would wait for up to 24m30.

**Comment by ncw** [comment]
Here is the data [in a google sheet](https://docs.google.com/spreadsheets/d/1Yd9mAdmxuOrWXeg91-porKYl0Cywy5-j_XNVN4jkVeQ/edit?usp=sharing) in case anyone wants to take a look

**Comment by felixbuenemann** [comment]
Nice stats! 125 MByte/s is roughly 1 GBps so that's likely the network bandwidth of the instances.

**Comment by felixbuenemann** [comment]
Maybe rclone could measure the time from transferring the last byte to the error, be it the expected 408 or something like 504 and include that in the measurement instead of hardcoding the 60 seconds, because otherwise rclone might wait too short if the error occurs shortly after the upload finished.

**Comment by left1000** [comment]
I'm concerned about a minor point you just made, your account is limited to 20 GB files? how is that possible? is this something that could happen to others? I know I have some 30gb files on my cloud, or at least I uploaded them fine, and they claim to still be there?

**Comment by ncw** [comment]
@felixbuenemann yes the 1gbps figure occurred to me too.  I'll see if it is easy to calculate when we sent the last byte and use that if possible. 

@left1000 I have a total limit of 20gb non photo files on my account. Amazon have stopped doing that now - you can only get the unlimited version.

**Comment by arionl** [comment]
@ncw Let me know if you'd like me to do any more test uploads. Thanks for digging into this and the data you collected makes a lot of sense. Do you intend on adding this logic to rclone so that if it is using a ACD target it'll automatically adjust ```--acd-upload-wait-time```?

BTW, I'm currently testing 3 scenarios with my 35gb video file using ```--acd-upload-wait-time 15m```:

- ACD 'copy': **success**
- ACD+Crypt 'copy': **success** (```v1.33-63-gace1e21```)
- ACD+Crypt+FUSE: pending (```v1.33-64-gbc414b6```)

I'm currently rsync'ing the same file for that last scenario and will report back.

BTW, I'd be happy to chip in to get you an unlimited Amazon Cloud Drive account. Your work on rclone has been inspiring.

**Comment by felixbuenemann** [comment]
@ncw I've added [some more values](https://docs.google.com/spreadsheets/d/1impnBVsvYP3punxTVsmg12Q610sQRh3iia8RjTDUj_Q/edit?usp=sharing) around the max filesize (49GB and 50GB never appeared). It seems the 30s/GB value should work reasonably well.

**Comment by arionl** [comment]
FYI, my third scenario in the [comment above](#issuecomment-253998322) failed. Attached is a log file. It looks like the file actually was transmitted and exists but isn't visible in the FUSE mount. If I do a normal listing, I can see it:

```
# rclone --config /opt/rclone/rclone.conf lsl acdcrypt:media
35190116352 2016-10-15 11:50:07.898000000 video.m2ts
2016/10/16 12:11:18
Transferred:      0 Bytes (0 Bytes/s)
Errors:                 0
Checks:                 1
Transferred:            0
Elapsed time:        1.5s
```

Doing a listing on the FUSE mount after this failed operation shows nothing:

```
# ls -al /mnt/acdcrypt/media/
total 0
```

However, if I unmount the FUSE path and then remount it, I can see the file but it looks like it wasn't 'finalized':

```
# ls -al /mnt/acdcrypt/media/
total 0
-rw-r--r-- 1 root root 35190116352 Oct 15 14:50 .video.m2ts.zyDwsj
```

Not sure if this is related to this specific issue or should be something moved to a new issue that is related to FUSE and/or crypt..

[acd-crypt-fuse-error.txt](https://github.com/ncw/rclone/files/531942/acd-crypt-fuse-error.txt)

**Comment by ncw** [comment]
OK here is my latest attempt to fix this problem.  See below for changed docs.

http://pub.rclone.org/v1.33-66-g2e20a0f-acd-timeout%CE%B2/

Any feedback very welcome, with suggestions for the default parameter values.  In my tests I successfully uploaded a range of file sizes up to 15 GB.  The tests weren't successful before I introduced the extra 2m sleep.

#### --acd-upload-wait-time=TIME, --acd-upload-wait-per-gb=TIME, --acd-upload-wait-limit=TIME ####

Sometimes Amazon Drive gives an error when a file has been fully
uploaded but the file appears anyway after a little while.  This
happens sometimes for files over 1GB in size and nearly every time for
files bigger than 10GB. These parameters control the time rclone waits
for the file to appear.

If the upload took less than `--acd-upload-wait-limit` (default 60s),
then we go ahead an upload it again as that will be quicker.

We wait `--acd-upload-wait-time` (default 2m) for the file to appear,
with an additional `--acd-upload-wait-per-gb` (default 30s) per GB of
the uploaded file.

These values were determined empirically by observing lots of uploads
of big files for a range of file sizes.

Upload with the `-v` flag to see more info about what rclone is doing
in this situation.

**Comment by ncw** [comment]
@arionl This issue caused the upload to fail most likely and rsync not to rename it to its final destination.  It didn't appear because of #680 most likely.  Note that acd doesn't support move yet so that is likely why it didn't get its final name from rsync #721.

**Comment by DurvalMenezes** [comment]
@ncw:
> rsync not to rename it to its final destination

You meant rclone, right?

**Comment by arionl** [comment]
@ncw ah, thanks for the clarification. I'll try again with a straight ```cp``` to see if it works (should be fine). I'll track the other issues for move/rename support.

**Comment by ncw** [comment]
I've merged this to master here: http://beta.rclone.org/v1.33-76-g5b83270/ (will be uploaded in 15-30 mins)

Let me know if you find any problems and I'll re-open the ticket if you do!

**Comment by left1000** [comment]
I just got used to splitting my very large files into parts. Glad someone finally solved this mystery. Thanks. Sorry if this thank you doesn't sage.

**Comment by felixbuenemann** [comment]
@ncw It seems there's a lot more fluctuation for Amazon Drive in the eu-west-1 AWS region. I've [updated the spreadsheet](https://docs.google.com/spreadsheets/d/1impnBVsvYP3punxTVsmg12Q610sQRh3iia8RjTDUj_Q/edit?usp=sharing) with the results of copying around 700 GB from an US amazon drive to an EU amazon drive with `--acd-upload-wait-per-gb 120s` and even with that setting, there were some timeouts that appeared after 0 seconds on re-upload, suggesting that they took longer than 120s per GB and there's one result that only made it because of the initial 60s delay with around 170s/GB.

**So maybe the default should be lowered to 3 minutes per gigabyte?**

Btw. rclone performance is impressive. I cloned my US to my EU drive at around 70 MB/s with 8 connections limited mostly by the max combined bandwidth of 1.5 GBps of the m1.large instance.

I'll continue copying with very high timeout values to get a better view of the worst case performance.

**Comment by ncw** [comment]
@felixbuenemann  I attempted to share [the doc](https://docs.google.com/spreadsheets/d/18wRo2JLojmXcuUUcbyco8RI5LEISNhNyOQI3QOPc6Ec/edit?usp=sharing) with you so you can edit it too. I've added your data to mine, and drawn a coulple of lines on the graphs with the current scheme and a proposed new scheme - feel free to adjust those settings.  It also shows the percentage hit rates of the data points we've got so far.

We could do with some more data with the bigger files.

The defaults currently stand at

```
	--acd-upload-wait-limit 60s
	--acd-upload-wait-time 120s
	--acd-upload-wait-per-gb 30s
```

Changing it to `--acd-upload-wait-per-gb 180s` would mean that a 50 GB upload (the max) might wait for 2.5 hours for it to appear.  Maybe that is OK though since a 50 GB upload likely took several hours to upload...

I'll all ears to your suggestions!

**Comment by felixbuenemann** [comment]
I'd like to play around with the `GET /resume` endpoint to see if it can be used as a more reliable way to get the upload status. because it allows to detect that an upload has failed or is still being checked. See the discussion in #87 for details. I barely know go, but I should be able to extend [acd_cli](https://github.com/yadayada/acd_cli) to interact with the endpoint.

**Comment by ncw** [comment]
@felixbuenemann interesting work!  I'd like to release the long delayed v1.34 pretty much as-is by the weekend so if you've got any thoughts on tweaking the existing timers for that, I'd be interested.  I suspect reverse engineering the resume might take a bit longer so let's do that in v1.35.

**Comment by felixbuenemann** [comment]
I've been busy at work last week, but collected more stats and also did some more tests with files around the 50 GB border. I'll add the numbers into the sheet later today.

**Comment by Originn** [comment]
I am uploading about 10 files of 48G each. I have created a batch file that will upload them automatically when the last upload had been verified. The waiting time is set to 20 minute between each file. Verification time can fluctuate from 5-18, but haven't reached 20 minutes yet. Is it possible to "shut off" waiting time, and allow the batch file to run without waiting for verification ?

**Comment by felixbuenemann** [comment]
@Originn You could try setting `--acd-upload-wait-time 1s --acd-upload-wait-per-gb 1s --retries 1` which should have the effect of exiting after the 60 second timeout by the acd api. I don't think you can avoid the initial 60 second wait for the answer to the PUT/POST request.

**Comment by Originn** [comment]
Thanks @felixbuenemann I have implement the command line in the batch file. The verification process can sometime take more time then uploading the file. I could retry uploading again,  when I see that it failed to upload a specific file out of the 10. It will be interesting to see which file it fails to upload.

**Comment by felixbuenemann** [comment]
@ncw I have updated the spreadsheet. I'm not sure why we need `--acd-upload-wait-time` if the value for `--acd-upload-wait-per-gb` is sufficiently large. Is the constant wait there to allow for some time for very small files to appear after let's say a 429 rate limited error?

I have played around with the "Limit2" values and 180s/GB is the only one that gives 100% success rate over the recorded test set (which we know already excludes some timeouts).

Personally I'd prefer waiting longer than needed over retrying, but you'll have to decide on a good compromise. If I'm on a slow uplink I would want to avoid the long time for the re-upload and if I'm on a fast uplink I would want to avoid the additional traffic caused by unnecessary retries, because fast links are usually metered.

It would be great if the verbose output always logged the error code that caused the waiting to be triggered, so we could get a better idea which status codes are always fatal. I have a suspicion that 400 and 500 error are always fatal while 408 and 504 are ok and 429 can be either failed or appear after some time, because the 429 (rate limited) error likely masks other response codes. Currently the original error code is only logged if the file never appeared.

**Comment by Originn** [comment]
@felixbuenemann When running this flag  --acd-upload-wait-time 1s --acd-upload-wait-per-gb 1s --retries 1

It does not recognize this specific flag acd-upload-wait-per-gb 1s

**Comment by felixbuenemann** [comment]
@Originn It is only implemented in the beta, you can omit it in the stable version.

**Comment by Originn** [comment]
@felixbuenemann Should look like this ? --acd-upload-wait-time 1s --retries 1

**Comment by felixbuenemann** [comment]
@Originn Yes.

**Comment by Originn** [comment]
@felixbuenemann Thank you, but this still doesn't work. Is there any other solution ?

**Comment by ncw** [comment]
Thanks for your analysis @felixbuenemann 

As a result I've drastically simplified the waiting into a single parameter 

    --acd-upload-wait-per-gb duration   Additional time per GB to wait after a failed complete upload to see if it appears. (default 3m0s)

Which is much more satisfactory.

See commit e162377ca36b5da680876ccb361892c3586a56ab

Can you have a go with the beta and report back please?  I'd like to make the v1.34 release tomorrow if possible!

http://beta.rclone.org/v1.33-106-ge162377/ (uploaded in 15-30 mins)

**Comment by felixbuenemann** [comment]
@ncw I switched to the new beta, not sure I'll have meaningful data by tomorrow though.

**Comment by felixbuenemann** [comment]
@ncw I accidentally mirrored to the wrong target folder between two cloud drives, so I ended up transferring 1.5 TB instead of only changes, but I'll likely not be able to analyze the logs today.

**Comment by ncw** [comment]
@felixbuenemann no worries.  I'm pretty sure the feature is working correctly so I've released it in v1.34 - if there is a problem then I'll fix it and make a new release.

**Comment by felixbuenemann** [comment]
@ncw Great, I've done some quick grepping over the logs and it appear to work fine.

Do you think you could add logging of the original error in the "Object found" log line, just like in the "Giving up waiting" line? This would make it easier to fine tune which errors are fatal and which ones should cause waiting in the future.
